### PR TITLE
fix: keep the newline char if file ends with that

### PR DESCRIPTION
**Pull request by s12f**
I found if the input file ends with the '\n', tlauc will delete the newline char, I think tlauc should keep the file same as possible, files ends with the newline char are quite normal(as https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline said, the Posix standard defines a line should end with a newline char, and some editors(e.g. neovim) will automatically put a newline char in the end).

### CODE REVIEW

**Comment by ahelwer** [comment]
Thanks! Could you add some tests for this? In particular, I would like to see the behavior if empty input is provided.

**Comment by s12f** [comment]
@ahelwer added some tests, in the cases that input is empty or empty ends with a newline, the `replace_symbols` will panic in origin codes, so I add https://github.com/tlaplus-community/tlauc/pull/20/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R457-R459

File: src/lib.rs [hunk 1]
```diff
@@ -850,4 +875,44 @@ op == /\ A
 ===="#,
         );
     }
+
+    // Tests that file ends with newline(or without newline)
```
Comment (ahelwer) [inline src/lib.rs]: nit: space between newline and (

File: src/lib.rs [hunk 2]
```diff
@@ -30,6 +30,9 @@ pub enum TlaError {
 }
 
 pub fn rewrite(input: &str, mode: &Mode, force: bool) -> Result<String, TlaError> {
+    // if the input ends with '\n', we should put the '\n' back to output
```
Comment (ahelwer) [inline src/lib.rs]: Can these lines be moved down below where they are used?

File: src/lib.rs [hunk 3]
```diff
@@ -432,6 +454,9 @@ fn mark_symbols(tree: &Tree, cursor: &mut QueryCursor, tla_lines: &mut [TlaLine]
 }
 
 fn replace_symbols(tla_lines: &mut [TlaLine]) {
+    if tla_lines.is_empty() {
+        return;
+    }
     for line_number in 0..tla_lines.len() - 1 {
```
Comment (ahelwer) [inline src/lib.rs]: Instead of special-casing an empty input array, change this to:
```suggestion
    for line_number in 0..tla_lines.len().saturating_add_signed(-1) {
```

File: src/lib.rs [hunk 4]
```diff
@@ -30,6 +30,9 @@ pub enum TlaError {
 }
 
 pub fn rewrite(input: &str, mode: &Mode, force: bool) -> Result<String, TlaError> {
+    // if the input ends with '\n', we should put the '\n' back to output
+    let end_of_newline = input.chars().last().map_or(false, |x| x == '\n');
```
Comment (ahelwer) [inline src/lib.rs]: A better variable name might be:
```suggestion
    let input_ends_in_newline = input.chars().last().map_or(false, |x| x == '\n');
```

File: src/lib.rs [hunk 5]
```diff
@@ -54,6 +57,25 @@ pub fn rewrite(input: &str, mode: &Mode, force: bool) -> Result<String, TlaError
     //println!("{:#?}", tla_lines);
     replace_symbols(&mut tla_lines);
 
+    // push a empty TlaLine,
+    // tla_lines...join("\n") will add a '\n' into end of the output,
+    // the reason why we don't push '\n' into output string directly is
+    // that maybe cause a reallocation.
+    if end_of_newline {
```
Comment (ahelwer) [inline src/lib.rs]: Move this functionality into the `Tla_lines::construct_from` method

**Review by ahelwer — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 5 above)
Comment (s12f) [reply src/lib.rs]: > Move this functionality into the `Tla_lines::construct_from` method

After re-thinking about that, I think changing `tla_lines` for output is a bad idea, if we put the functionality into `Tla_lines::construct_from`, that will be very confused, for examples, the input `"abc\n"` is exact one line by definition, so `Tla_lines::construct_from` should return `["abc"]` instead of `["abc", ""]`.

The newline problem is caused by the output logic, so I write the `TlaLine::output_from_lines` to handle that.

**Comment by ahelwer** [comment]
Great changes, thanks!
